File: 0876-hand-of-straights/0876-hand-of-straights.cpp
class Solution {
public:
    bool isNStraightHand(vector<int>& hand, int groupSize) {
        int n=hand.size();
        if(n % groupSize !=0) return false;
        unordered_map<int,int> mpp;
        for(int i=0;i<n;i++){
            mpp[hand[i]]++;
        }
        for(int it : hand){
            int start = it;
            while(mpp[start-1]) start--;
            while(start <= it){
                while(mpp[start]){
                    for(int next = start; next< start+ groupSize;next++){
                        if(!mpp[next]) return false;
                        mpp[next]--;
                    }
                }
                start++;
            }
        }
        return true;
    }
};

File: 0876-hand-of-straights/0876-hand-of-straights.java
class Solution {
    public boolean isNStraightHand(int[] hand, int groupSize) {
        Map<Integer, Integer> map = new HashMap<>();
        int n = hand.length;
        if(n%groupSize != 0) return false;
        for(int num:hand) map.put(num, map.getOrDefault(num,0) + 1);
        Arrays.sort(hand);
        
        for(int num:hand) {
            if(map.get(num) == 0) continue;
            for(int i=0;i<groupSize; i++) {
                if(!map.containsKey(num+i) || map.get(num+i) == 0) return false;
                map.put(num+i, map.get(num+i) - 1);
            }
        }
        return true;
    }
}